### ISSUE TITLE
Stop using / as division

### ISSUE THREAD

**Issue by nex3** [sass/sass#2565]
**The current plan:**

1. Create an automated migration tool that can migrate users from `/`-as-division to a `divide()` function.
2. Deprecate `/`-as-division. Any time a `/` would be interpreted unambiguously as division, or an ambiguous `/` would be made unambiguous (for example by being stored in a variable), we produce a deprecation warning .
3. Add a top-level `divide()` function (`math.div()` in the new module system). This will work exactly the same way division works today, with the obvious exception that it will never be treated as a slash separator.
4. Add a new slash-separated list type to represent slash as a separator. For the time being, this will not have a literal syntax.
5. Add a `slash-list()` function (`list.slash()` in the new module system) that returns a slash-separated list. Providing the ability to work with slash-separated lists even before they have a literal syntax will help stylesheets be forwards-compatible with the new wold.
6. Wait until at least three months after the module system is released.
7. Release Dart Sass 2.0.0, which parses `/` as a list separator that binds more loosely than space but more tightly than comma.
8. Add support for [first class `calc()`](https://github.com/sass/sass/issues/2186), which will allow `/`-as-division in the form of `calc(5px / 2)`.
8. Deprecate the `slash-list()`/`list.slash()` function in favor of using the literal syntax.
9. Whenever Dart Sass 3.0.0 releases, remove `slash-list()` and `list.slash()`.

* [x] [Migrator](https://github.com/sass/migrator/issues/37)
* [x] [Proposal](https://github.com/sass/sass/blob/master/accepted/slash-separator.md)
* Deprecation, `divide()`, and `slash-list()`
  * [x] [Specs](https://github.com/sass/sass-spec/pull/1399)
  * [x] [Dart Sass](https://github.com/sass/dart-sass/issues/663)
  * ~[LibSass](https://github.com/sass/libsass/issues/2887)~
* Removal of /-as-division, deprecation of `slash-list()`
  * [x] Update deprecations (https://github.com/sass/sass/pull/4131)
  * [x] Specs (https://github.com/sass/sass-spec/pull/2086)
  * [x] [Dart Sass](https://github.com/sass/dart-sass/issues/663) (https://github.com/sass/dart-sass/pull/2649)
  * [ ] Sass Site
  * ~[LibSass](https://github.com/sass/libsass/issues/2887)~
* Removal of `slash-list()`
  * [ ] Update deprecations
  * [ ] Specs
  * [ ] [Dart Sass](https://github.com/sass/dart-sass/issues/663)
  * ~[LibSass](https://github.com/sass/libsass/issues/2887)~

---

**The original issue:**

When Sass was first designed, the most prominent use of `/` as a separator in CSS was the short-form [`font` property](https://developer.mozilla.org/en-US/docs/Web/CSS/font), which used it to separate values for `font-size` and `line-height`. Because this was a relatively narrow edge case, we were able to use `/` as the division operator as well, with [some heuristics](https://sass-lang.com/documentation/file.SASS_REFERENCE.html#division-and-slash) to decide between them. When `/` was used as a separator, the value it produced was actually the divided number, but it was *rendered* with the slash instead.

However, the slash separator has been seeing more use in recent CSS specs. The [`grid-row`](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-row) property uses it to separate values for `grid-row-start` and `grid-row-end`, and [Colors Level 4](https://drafts.csswg.org/css-color/) uses it in many color functions (including the new recommended syntax for `rgb()` and `hsl()`) to separate the alpha value from the rest of the parameters.

The increased prominence of this syntax means that the sharp edges of the heuristics we use to distinguish between `/`-as-separator and `/`-as-division will start cutting more users. Once Sass supports the new color function syntax (https://github.com/sass/sass/issues/2564), users are likely to try passing the alpha by variable, which will currently fail. We need to decide whether to do something about this, and if so, what.

## How do we represent division?

Today, division is represented by the operator `/`, which as discussed above is ambiguous. We could fairly easily add a new syntax for this, which would allow us to migrate users away from `/`, eventually leaving it as unambiguously a separator.

I see a few options for this:

* Keep the behavior as-is. This has the substantial benefit of requiring no migration, and it will be sufficient to support compatibility with plain CSS. However, it will make it much more painful for users to use dynamic alpha values for their colors as the new color function syntax gains traction.

* Replace the `/` operator with another operator. There's no option that's really consistent with other programming languages, but `~` is a possibility that's currently unused and looks a *little* like ÷.
https://github.com/sass/sass/issues/2565
* Replace the `/` operator with a function, such as `div()`. This is more verbose, but it's very clear to the reader and it avoids any risk of further collision with new CSS syntax.

## What is a slash separator?

If we do decide to free up `/` to unambiguously represent a separator, we need to figure out how that separator works in terms of SassScript. This is particularly tricky because its scope varies from property to property. In `font`, it's *tightly bound*--the slash in `italic small-caps bold 16px/2 cursive` separates `16px` from `2`, but it *doesn't* separate `italic small-caps bold 16px` from `2 cursive`. However, in `grid-row` and the color functions, it's *loosely bound*--the slash in `rgb(1 2 3 / 0.5)` separates `1 2 3` from `0.5`, and is not bound to `3` in particular.

I think it makes sense in general to consider `/` another kind of list separator, so that it works smoothly with existing APIs. But because of the scoping differences it's not clear whether it should bind more tightly than space-separated lists or not. That is, should `1 2 / 3 4` be equivalent to `(1 2) / (3 4)` or `1 (2 / 3) 4`?

**Comment by nex3** [comment]
@tabatkins Any insight you can provide would be welcome. In particular:

* How likely is the `/` separator to continue to be used in new rules/functions?

* If `/` is used more, is it likely to bind tightly or loosely?

* How likely is `~` to appear as syntax in a property value?

**Comment by jelbourn** [comment]
Would it make sense to continue interpreting `/` as division in places where it would otherwise be invalid CSS, otherwise treating it as the native separator? That would necessitate introducing a division function, but it would only need to be used when there would be ambiguity. The compiler could error on any ambiguity.

Admittedly, this _could_ get confusing. It might be valuable to analyze existing libraries to see how prevalent ambiguities would actually be.

**Comment by nex3** [comment]
Sass doesn't have any contextual information about the property which it's evaluating code for. This is both because we have a general policy of encoding as little information about CSS itself as possible, and because there's no way to know what property a given variable or argument might be used with. Without that, I'm not sure there's a way to know which uses of `/` would be valid CSS and which wouldn't.

**Comment by rickcnagy** [comment]
Is there any reason that `//` isn’t an option? It would be similar to Python 3.

**Comment by nex3** [comment]
`//` is already the syntax for a single-line/silent comment.

**Comment by tabatkins** [comment]
> How likely is the / separator to continue to be used in new rules/functions?

We will continue using it rarely, as we do today, as a secondary separator.

> If / is used more, is it likely to bind tightly or loosely?

The general rule, which we *almost always* stick to, is that CSS separates repetitions of values by `,`, then within each repetition (or within the whole value, for non-repeating properties), primarily uses ` ` (space) to separate tokens, only using `/` as a separator when required due to ambiguity.  We don't have a strict notion of binding strength for the `/`; the only firm rule is that `,` binds more loosely than space or `/`.

(More precisely, *sometimes* `/` binds tightly to the next token, like in `font` or `rgb()`, while in others it binds loosely and separates groups of space-separated tokens, like in `border-radius` or `border-image`.)

> How likely is ~ to appear as syntax in a property value?

Beyond it's current use in selectors, I don't *expect* it to show up property values; we tend to be conservative about new characters to reduce grawlix-ness. That said, it's still possible that we add a `selector()` function or similar at some point in the future, allowing `~` to show up at the property level. (Selectors will *not* show up "unshielded"  by a function tho, I'm sure of that; their syntax is too open-ended to reasonably do that.)

That said, the fact that it *is* already used in Selectors *does* reduce the friction for using it some time in the future; other proposals for using new ASCII characters sometimes founder on "that's easy to type on an English keyboard, but not on my [other-language] keyboard", but ~ is already a sunk cost in that regard.

We did vacate the naked-parenthesis space for y'all, and I'm willing to defend that line, so if you want to reserve "within ()" as the space for safe free-form math, I think you can safely do so.  Or, of course, just use "calc()". ^_^

**Comment by nex3** [comment]
> The general rule, which we *almost always* stick to, is that CSS separates repetitions of values by `,`, then within each repetition (or within the whole value, for non-repeating properties), primarily uses ` ` (space) to separate tokens, only using `/` as a separator when required due to ambiguity.

Out of curiosity, why did you end up with `rgb(r g b / a)` rather than `rgb(r g b a)`? It seems like the latter shouldn't be ambiguous.

> (More precisely, *sometimes* `/` binds tightly to the next token, like in `font` or `rgb()`, while in others it binds loosely and separates groups of space-separated tokens, like in `border-radius` or `border-image`.)

I actually hadn't realized that `border-radius` and `border-image` had `/` as part of their value grammars. In that case, binding looser than space *definitely* hits the most properties today, and there's no reason to think that won't hold true in the future. It makes the parse tree for `font` pretty weird, but it will still render fine.

> That said, it's still possible that we add a `selector()` function or similar at some point in the future, allowing `~` to show up at the property level. (Selectors will *not* show up "unshielded"  by a function tho, I'm sure of that; their syntax is too open-ended to reasonably do that.)

Anything shielded in a function name is totally fine for us syntax-wise. We have a strong precedent with `calc()` for having special parsing for certain functions.

> That said, the fact that it *is* already used in Selectors *does* reduce the friction for using it some time in the future; other proposals for using new ASCII characters sometimes founder on "that's easy to type on an English keyboard, but not on my [other-language] keyboard", but ~ is already a sunk cost in that regard.

That also makes it appealing for us. On the whole, it sounds like `~` is a likely to be our best option if we do decide to continue to use a single character.

> We did vacate the naked-parenthesis space for y'all, and I'm willing to defend that line, so if you want to reserve "within ()" as the space for safe free-form math, I think you can safely do so.  Or, of course, just use "calc()". ^_^

Leaving parentheses alone has been extremely helpful :bowing_woman:, and in fact it is one of the heuristics we use today to decide what to do with `/`. But as discussed above, those heuristics are likely to become more annoying the more prevalent slash-separation becomes.

**Comment by tabatkins** [comment]
> Out of curiosity, why did you end up with rgb(r g b / a) rather than rgb(r g b a)? It seems like the latter shouldn't be ambiguous.

It's not *technically* ambiguous in `rgb()` and `hsl()`, but the `/` does serve as a nicely visible indicator that "hey, this color has an explicit alpha": in `rgb(0 0 0 0)` vs `rgb(0 0 0)` vs `rgb(0 0 0 / 0)`, the first two aren't super-apparent on first glance, but the last two are really obvious.

But we were also trying to solve the problem for `color()`, which can take any number of arguments after the colorspace name, and we wanted to enable some of those to be optional. There's no easy way to add an easily-readable, easily-writeable, unambiguous alpha value there, without relying on some form of separator, and we already wanted to use the `,` for its normal "separating repetitions" functionality, so you could supply a fallback color. `/` ends up serving its "disambiguate a value from similar surrounding values" function really well there, so we applied it globally to the other colors.

Finally, as a minor but significant additional point, `gray()` only takes a single color-relevant argument, so with alpha it has two values. Between rgb/hsl/lab/lch all taking 3+alpha, color taking n+alpha, and gray taking 1+alpha, having an explicit visual indicator of when the alpha was being set makes things more readable in general, without having to remember how many arguments any particular function takes before the alpha.

> Anything shielded in a function name is totally fine for us syntax-wise. We have a strong precedent with calc() for having special parsing for certain functions.

Cool. We'll never have a selector more complicated than an ID unshielded. (I know that still offers some ambiguity for you, as both IDs and hex colors use the same token, but eh. *Most* of the time it's still unambiguous due to character sets. ^_^)

**Comment by nex3** [comment]
I discussed this with a crack team of Sass users at Google, and we came up with the following plan:

1. Deprecate `/`-as-division. Any time a `/` would be interpreted unambiguously as division, or an ambiguous `/` would be made unambiguous (for example by being stored in a variable), we produce a deprecation warning .
2. Add a top-level `divide()` function (`math.div()` in the new module system). This will work exactly the same way division works today, with the obvious exception that it will never be treated as a slash separator.
3. Add a new slash-separated list type to represent slash as a separator. For the time being, this will not have a literal syntax.
4. Add a `slash-list()` function (`list.slash()` in the new module system) that returns a slash-separated list. Providing the ability to work with slash-separated lists even before they have a literal syntax will help stylesheets be forwards-compatible with the new wold.
5. **After a suitable deprecation period,** start parsing `/` as a list separator that binds more loosely than space but more tightly than comma.
6. Deprecate the `slash-list()`/`list.slash()` function in favor of using the literal syntax.
7. In the next major revision, remove `slash-list()` and `list.slash()`.

This is not set in stone. It's a big change, and I want to make sure people have a chance to provide feedback before I start working on a more formal spec. I'll leave this open for at least a few weeks so folks have a chance to chime in.

One major open question (at least for Dart Sass) is, how long do we leave the deprecation before we drop support for `/` as division, and do we do a major version bump at that point? Because this is a CSS compatibility issue, [Dart Sass's compatibility policy](https://github.com/sass/dart-sass#compatibility-policy) gives us license to remove the old behavior as early as three months after the deprecation warning is released, without a major version bump. However, this affects a much broader range of stylesheets than most, so I think that's too short. Would six months but no major version bump work?

**Comment by fedfigca** [comment]
My ¢2 is that / is commonly used as division almost everywhere else, using a different syntax will be yet another challenge for new users, and for old code alike, so trying to keep it should be the priority.

Using some sort of encapsulation for the arithmetic might be one way to proceed in places where the slash could mean a separator, or any other way, but if we have to teach new users (who probably already program other languages) a different syntax for math, it might become one of those issues that make people look for alternatives.

**Comment by weo3dev** [comment]
> is that / is commonly used as division almost everywhere else

Guess what. CSS is not "everywhere else", and, just like any programming language, CSS has a designed syntax agreed upon not just by open source organizations but also the platforms in which we see CSS in action.

@fedfigca *All* CSS developers need to [came across too strongly here] _get up to speed_ and come to recognize the paradigm shift that is CSS Grid. It was built specifically to fix a two-decades-old problem in the CSS spec of there not being any one tool specifically for layout. We have used hacks and workarounds for over 20 years because we never had CSS grid from the outset.

Now we do, and part of its designated and designed spec is using the forward-slash as a means of defining columns and rows, and spans of columns and rows for any grid child.

Any NEW user of CSS should be learning that for _layout_, we now use CSS Grid. Period. Full stop. 

Saying, "so trying to keep it [forward slash as division only in Sass] should be the priority" is really ignoring the larger picture of the progression of the language Sass is built to parse.

**Comment by weo3dev** [comment]
@nex3 I do not envy you, and wish you the best of luck in moving forward with the planned roadmap.

**Comment by fedfigca** [comment]
> 
> 
> > is that / is commonly used as division almost everywhere else
> 
> Guess what. CSS is not "everywhere else", and, just like any programming language, CSS has a designed syntax agreed upon not just by open source organizations but also the platforms in which we see CSS in action.

No need to "guess what" anybody here, we're just talking and pitching in our points and opinions, if you come at people like that you'll scare off valuable opinions 😉

I still think that doing math differently is a mistake, is not my decision to make, it'll be what it'll be, and people will use whatever they find the most comfortable and less painful.

**Comment by nex3** [comment]
> Using some sort of encapsulation for the arithmetic might be one way to proceed in places where the slash could mean a separator, or any other way, but if we have to teach new users (who probably already program other languages) a different syntax for math, it might become one of those issues that make people look for alternatives.

We do eventually plan to have Sass natively support `calc()`: https://github.com/sass/sass/issues/2186. Once that lands, if you do a calculation that's fully computable at compile-time, it will return a number, so you'll be able to write `calc($width / $height)` instead of `divide($width, $height)`.

There are some very real downsides to making that *the only* way division works, though. First, encouraging users to use `calc()` for local calculations can hide errors. Rather than failing, we'd probably produce a plain-CSS `calc()` for any unit math Sass doesn't understand, so if someone tries to write something like `calc(1/2em)` they'll end up with a broken property that's hard to debug. Having a Sass-only way to divide is a good way of getting users to express when they expect an operation to be doable by Sass, so Sass can fail when it's not doable.

Second, it means we can't solve this issue until #2186 is completely solved. First-class calc is a very large and complex feature, and designing, speccing, and implementing it will take a long time—especially since we'd need to wait until LibSass supports it to even begin deprecating `/`-as-division. I don't want to speak for @xzyfer, but I'd guess that LibSass's development time is going to go almost entirely towards the module system for the next six months at least, so that means it may be as much as a year before we could even begin to address this issue.

**Comment by fedfigca** [comment]
> There are some very real downsides to making that _the only_ way division works, though. First, encouraging users to use `calc()` for local calculations can hide errors. Rather than failing, we'd probably produce a plain-CSS `calc()` for any unit math Sass doesn't understand, so if someone tries to write something like `calc(1/2em)` they'll end up with a broken property that's hard to debug. Having a Sass-only way to divide is a good way of getting users to express when they expect an operation to be doable by Sass, so Sass can fail when it's not doable.
> 

Your points about doing it all with calc() are very true, doing something different like math() (or any better name you can think of) and allowing that to take the usual math syntax in a similar way of what calc takes might be a better way than to have something to just divide, and letting Sass know that it should take over the calculations instead of producing calc() values

**Comment by Cleecanth** [comment]
I'm sure I'm leaving out a ton of complexity, but I second the idea of using a "fenced"/psuedo-function for Sass-specific math (something like `math(2/3)`).

It feels cleaner than eliminating a common character for math functions and it makes for a pretty straight-forward upgrade path. 

Having written some fairly complex Sass over the past 6-or-so years, I've tended to gravitate more and more towards clarity over convenience, though, so I'm a bit biased. At the same time, a wrapper around Sass-specific math still feels like it maintains the spirit of both languages (CSS and Sass) while still being pretty basic to write.

I also realize this isn't one of the original proposals, so it's very possible there's a good reason not to do it. In which case, I could probably get behind using `~` (or really any other character[s]).

**Comment by nex3** [comment]
Something like `math()` definitely seems tempting once you understand all the ins and outs of the problem, but I suspect for a user who's not super familiar with Sass (and possibly not with CSS either), it's likely to be very confusing. It's not obvious from seeing `math()` how it's different than `calc()` (especially once `calc()` supports a superset of its behavior), why you need to use it for division and not other arithmetic operations, or even whether it's a Sass feature or a plain CSS feature.

It occurs to me that it may be possible to mitigate (although probably not eliminate) the set of invalid `calc()`s that could be generated. CSS [clearly defines](https://www.w3.org/TR/css-values-4/#calc-type-checking) what it means for a `calc()`'s type to be valid or not, so we can eagerly fail if someone attempts to write `calc(1/2em)`. This leads me to believe that the best long-term solution for writing division using `/` is to do it within `calc()`.

**Comment by nex3** [comment]
I've brought everyone's concerns to my user team, and we've affirmed that we want to proceed using `divide()` without waiting for `calc()` support or adding an interim `math()` syntax. However, we *will* wait to enable with the deprecation until we have a migration tool in place to make it easy for users to update their stylesheets.

We've also decided on a more concrete timeline. We'll launch the deprecation as soon as possible after the migration tool is implemented, then land the removal in a major version release no earlier than three months after the launch of the module system. We want to ensure a healthy overlap of module-system-supporting versions and division-operator-supporting versions so users can opt to migrate to the module system without migrating away from the division operator.

**Comment by weo3dev** [comment]
Hey @fedfigca - apologies for not responding earlier - thank you for calling me out on my attitude that is not a benefit to the community. Hoping the decision to move forward is a good one. Peace.

**Comment by nex3** [comment]
As a small adjustment, I think we should also make the `divide()` function accept string arguments and do the same thing the `/` operator does with strings now (create an unquoted string containing `/`). It should definitely emit a deprecation warning when a string is passed, but allowing strings will make automated migration much easier.

**Comment by nex3** [comment]
[The official proposal pull request](https://github.com/sass/sass/pull/2683) is now live for comments!

**Comment by mirisuzanne** [comment]
I understand that I'm late to this conversation, but I want to push back on a few things as an author:

One of Sass's long-term strengths has been the shallow learning-curve for people who do not consider themselves programmers. Another has been the ability to do math in CSS (especially before `calc()` was reliable, and still for math that should be precompiled). I'd worry that handling one specific math operator in a totally new way – unlike all other math operators in Sass – causes issues on both fronts. If that had been the case when I picked up Sass, I likely would have been scared away by the strange seemingly-arbitrary design of that feature. 

I'm also starting to question the long-term compiled-`calc()`proposal. While it's super elegant at first glance - it strikes me as a potential repeat of the `@import` issue (and `-` namespacing issue) – ignoring the general rule that Sass-compiled additions to CSS syntax should be clearly marked and explicit. PostCSS plugins have made a mess out of this several times (see attempts to precompile custom properties) – and it's always frustrating as an author to have my CSS code altered without clear ways to turn that off.

I think an explicit math function of some kind would actually be more clear on both fronts - short and long term. I might consider naming it something like `precalc()` to acknowledge the similarity, and highlight the difference between the two. For now, `prrecalc()` could simply provide a safe context for `/`-as-division, and down the road it could potentially take on the proposed smart-`calc()` role, while still being explicit.

**Comment by mirisuzanne** [comment]
The module system also helps clear up any confusion - since any `precalc()` function would have to be loaded explicitly from a Sass namespace, e.g. `math.precalc()` – where `calc()` and other sass functions are global.

**Comment by nex3** [comment]
> One of Sass's long-term strengths has been the shallow learning-curve for people who do not consider themselves programmers. Another has been the ability to do math in CSS (especially before `calc()` was reliable, and still for math that should be precompiled). I'd worry that handling one specific math operator in a totally new way – unlike all other math operators in Sass – causes issues on both fronts. If that had been the case when I picked up Sass, I likely would have been scared away by the strange seemingly-arbitrary design of that feature.

I don't think it's quite fair to say that `divide()` is unlike all other math operators in Sass—things like rounding, absolute values, and so on are all functions. Once we add exponentiation, that'll be a function as well, even though in some programming languages it's `**`.

I see what you're saying that division is most commonly represented in programming languages as `/`, and that many users will expect it to be there. I'd like for it to be there too, but we have to balance that out with the increasingly-painful problems with CSS treating `/` as a separator. And I don't think it's likely to be *as* upsetting as you do, especially if we can provide error messages that explain what's going on when it's likely that a user accidentally used `/`.

> I'm also starting to question the long-term compiled-`calc()`proposal. While it's super elegant at first glance - it strikes me as a potential repeat of the `@import` issue (and `-` namespacing issue) – ignoring the general rule that Sass-compiled additions to CSS syntax should be clearly marked and explicit. PostCSS plugins have made a mess out of this several times (see attempts to precompile custom properties) – and it's always frustrating as an author to have my CSS code altered without clear ways to turn that off.
> 
> I think an explicit math function of some kind would actually be more clear on both fronts - short and long term. I might consider naming it something like `precalc()` to acknowledge the similarity, and highlight the difference between the two. For now, `prrecalc()` could simply provide a safe context for `/`-as-division, and down the road it could potentially take on the proposed smart-`calc()` role, while still being explicit.

I'd say that first-class `calc()` will be more like Sass's support for CSS units and colors than like Sass's `@import`. We're just using `calc()` to mean the same thing it does in CSS, but with extra abilities. The fact that you can use variables in it is no different than using variables in a call to `rgb()`, and still clearly signals that something Sass-y is going on; and the fact that we can detect when it's completely resolvable at runtime and emit `2.5px` rather than `calc(5px/2)` is just an optimization, like converting `hsl()` to hex color codes.

**Comment by mirisuzanne** [comment]
> I'd say that first-class calc() will be more like Sass's support for CSS units and colors than like Sass's @import.

That's a great point. Put in that context, I see what you mean, and why `calc()` would work. We just have to avoid the over-optimizations that some tools attempt: like the minimizer that recently removed the unit from a `0%`, totally breaking my `hsl()`. I trust that.

I don't think I'm wrong to be somewhat worried about `divide()`, though. I realize _something_ has to be done, but I'm concerned with keeping it as CSS-like and readable as possible. Of the proposals mentioned, `divide()` feels the least CSS-like to me. In CSS, there is an established pattern for `/` as division, and for fencing math equations inside a function – but there is no pattern for basic math operations themselves being functions. `sass.math()` seems both more readable, and more CSS-like to me than `divide()`.

I'm not really thinking about how _most programming languages_ represent division. I actually use functional versions of `+ - * / =` in my code often – I don't think they're bad to have – but I  do think they stray farther from CSS in a very noticeable way for people unfamiliar with imperative languages. Unlike `pow()` or `abs()` or other advanced math functions, `+ - * / =` are pretty universal symbols outside any technical field. I can generally write those in an email to anyone I know, without any confusion. I'm specifically worried about how this would come across to people who don't work with other languages besides CSS.

**Comment by nex3** [comment]
> I don't think I'm wrong to be somewhat worried about `divide()`, though. I realize _something_ has to be done, but I'm concerned with keeping it as CSS-like and readable as possible. Of the proposals mentioned, `divide()` feels the least CSS-like to me. In CSS, there is an established pattern for `/` as division, and for fencing math equations inside a function – but there is no pattern for basic math operations themselves being functions.

This is becoming much less true. [Values and Units Level 4](https://drafts.csswg.org/css-values-4/#math) currently defines `min()`, `max()`, `clamp()`, `sin()`, `cos()`, `tan()`, `asin()`, `acos()`, `atan()`, `atan2()`, `pow()`, `sqrt()` and `hypot()`. Most of these aren't yet implemented in browsers, but we should design for the future where they probably will be.

> I'm not really thinking about how _most programming languages_ represent division. I actually use functional versions of `+ - * / =` in my code often – I don't think they're bad to have – but I do think they stray farther from CSS in a very noticeable way for people unfamiliar with imperative languages. Unlike `pow()` or `abs()` or other advanced math functions, `+ - * / =` are pretty universal symbols outside any technical field. I can generally write those in an email to anyone I know, without any confusion. I'm specifically worried about how this would come across to people who don't work with other languages besides CSS.

This is fair, but I'd expect users from a non-programming background to come in with *much* looser expectations for how math will be represented than people who have programmed before. Even given that, I don't think choosing a different operator (like `~`) is likely to be much friendlier than `math.div()`—there's no other character that has *any* mindshare for non-technical users as division, except I guess `÷` which is prohibitively difficult to type. Something like `precalc()` is also likely to trip novice users up, given that all other mathematical operations can be freely written outside of a function context.

I definitely see what you're saying about the downsides of not supporting `/`, but I don't see that the alternatives are better. Choosing a different operator character will make *reading* division harder as well as writing, and a Sass-only fence like `math()` or `precalc()` is impossible for users to guess and will provide very limited value once we support first-class `calc()`.

**Comment by mirisuzanne** [comment]
I think what I’m actually suggesting is that the most CSS-like and consistent is to fence *all math* inside some form of calc function, rather than trying to find the perfect way to  special-case division. It’s a bigger adjustment from Sass historically- but I think it might actually add clarity to both reading and writing the language. At least, I tend to try and fence math somehow anyway, for readability. If we’re gonna make _some_ breaking change - we could rethink our math solution entirely. Really, that’s what you are proposing with `calc()` as well - I’m just suggesting that a `math()` or `precalc()` gets us a lot closer in both syntax and concept than a `div()` special-case. Also a simpler manual upgrade path both times. 

Though I’m totally in favor of all math operations getting functional alternatives as well, now that we can namespace them :)

**Comment by nex3** [comment]
The backwards-incompatibility pain of deprecating *all* mathematical operators is almost certainly too severe to bear if it's not strictly necessary for CSS compatibility, even if it were something we thought would be a substantial readability improvement, which I'm honestly pretty suspicious of. I think many Sass users, both novice and invested, would disagree with your sense that explicitly bracketing all math is a clarity win.

**Comment by mirisuzanne** [comment]
Haha, that's totally fair. :) Deprecating all un-fenced math is clearly too severe, and I know nothing about actual difficulty programming these things in Dart. So one last question: is there a major downside here to providing the options:

- namespaced functions available for all math operators
- a way of fencing existing math syntax

Is that downside mostly in the time it would take to write that feature?

**Comment by nex3** [comment]
Adding more features always comes at a cost to the overall complexity of the language, and that's doubly true for multiple ways to do the same thing—it means users have to choose when to use one or the other, and readers familiar with one have to figure out why the other is used instead and if there's any material difference between them. Ultimately this usually ends in a style guide enforcing only one option anyway.

As far as fencing existing math syntax goes, `calc()` will be a way to do that.

**Comment by atg** [comment]
I would question whether the CSS grid syntax is actually more important than backwards compatibility (presumably the color syntax can be special cased).

I would also question whether Google is a representative sample of sass users. As far as I'm aware Google has all their code in a monorepo and pays engineers to perform these kinds of migrations. The proposed migration plan seems perfect for Google, but does it not seem perfect for the rest of us.

6 months is certainly too short for a transition period -- a change of this magnitude would take years. I don't know if I could even *locate* every sass compiler on my computer, let alone update them all regularly. Sass compilers are like glitter, they get everywhere - globally, as project dependencies, in CIs, in text editors.

The biggest problem is taking valid code, making it invalid for a period of time, and then making it valid again X amount of time later with totally different semantics. After the transition period is up it will be impossible to discern which code is pre-transition (MUST be migrated) and which is post-transition (must NOT be migrated).

**Comment by nex3** [comment]
Breaking changes certainly do suck, but in some cases they're unavoidable. One of the most important guarantees Sass makes is that it's a CSS superset, which means that plain CSS must be valid Sass *with the same meaning it has in CSS*. Sticking to this principle, even when it's meant painful migrations, is a big part of what's made Sass successful over the long term. We try to make decisions that will still be right five or ten years down the road.

The CSS working group's use of forward slash as a separator has increased considerably over the last few years, and it's a reasonable bet that that will continue. Maybe we could continue to special-case the color syntax (although I emphasize that it's *very* awkward to make that workaround happen), but we can't special-case everything. A transition needs to happen, even if it hurts.

**Comment by khamer** [comment]
I'm not sure if there's a better place to document this; let me know. One of the side effects is that projects that hadn't transitioned to `@use` from `@import` may run into issues. It's possible to create a shim _divide.scss, and you can import once globally and then use `divide()` instead of `math.div()` and avoid putting `@use` statements in other files yet:
```scss
@use "sass:math";

@function divide($a, $b) {
    @return math.div($a, $b);
}
```

I understand why you wouldn't want to add a new, unnamespaced function `divide()`, but this shim should avoid the warning for new projects without them needing to refactor anything from `@import` to `@use` yet.

**Comment by antoniandre** [comment]
Atm, the new `math.div()` syntax still breaks in vue-cli (I am also trying in Vite where it does not break), but we can't upgrade the sass-loader dependency yet since vue-cli runs on Webpack 4. Not sure if there's a solution yet.
But for the moment preparing my code to be compatible, I think @khamer's idea is great, in order to have that change remaining only in a single place for when it will be supported.
> 
> ```css-scss
> @use "sass:math";
> 
> @function divide($a, $b) {
>     @return math.div($a, $b);
> }
> ```

Also in a long term, I like the idea to not have to import `sass:math` at the top of all the files which use the new math.divide, especially in all the `.vue` components. :/
So in the end I think this divide mixin looks great. I feel it's also shorter to type and simpler to read than `math.div`!  👍

**Comment by nex3** [comment]
@khamer

> I'm not sure if there's a better place to document this; let me know. One of the side effects is that projects that hadn't transitioned to `@use` from `@import` may run into issues.

It's entirely possible to use `@use` in files that otherwise use `@import` (or are loaded with `@import`). The new module system was designed to be as interoperable with the old `@import` world as possible.

@antoniandre 

> Atm, the new `math.div()` syntax still breaks in vue-cli (I am also trying in Vite where it does not break), but we can't upgrade the sass-loader dependency yet since vue-cli runs on Webpack 4.

Breaks it how? Is there a tracking bug for this?

**Comment by moleCuleFFF** [comment]
This is rubbish slashes have been used in both operators and path delimiters in the majority of programming languages for years, the onus should be on the context to describe it's use correctly.

Why or how this BREAKING change was merged is beyond me, your breaking years of precedent and bulking perfectly readable syntax for nothing.

**Comment by nex3** [comment]
@moleCuleFFF Please remember the [community guidelines](https://sass-lang.com/community-guidelines): Disagreement is no excuse for poor manners.

Path delimiters are not part of the expression syntax of most programming languages. I know of no language other than Sass where the `/` in `1px / 2px` could be a separator or a division operator depending only on the context of the expression in question. The fact is that as long as CSS was going to expand its use of the `/` separator as it has over the last few years, our hands are tied: we needed to get rid of `/`-as-division. The only question is what the right *way* of doing so was, and the time for discussing that has come and gone.

**Comment by moleCuleFFF** [comment]
Forward slash is the most common used path delimiter in computing, https://en.m.wikipedia.org/wiki/Path_(computing)

It's also the most common operator for division, so with both these things being true I don't see ur point.

**Comment by moleCuleFFF** [comment]
@nex3 Can I ask who is the arbiter of "Poor Manners". Are we talking Eaton levels of queens english here ? I don't think my manners were bad, they were to the point, and strong, but not rude. I still think this committ is rubbish, it's my opinion and I am fully entitled to it, I in no wayt attacked you or your team personally, I think you are maybe a little too sensitive to be championing a change as divisive as this if you find me expressng an opinion as poor manners.

**Comment by lk77** [comment]
hello,

how to disable this warning ?

running the migrator against node_modules does not seems to be a solution to me

for me it does not make sense to display a warning at all,
just talk about it in the upgrade guide for v2, 
but stop spamming with deprecation notices when the end user don't care about, 
and can't do nothing about it because those deprecation notices are from node_modules.

**Comment by kerryj89** [comment]
> hello,
> 
> how to disable this warning ?
> 
> running the migrator against node_modules does not seems to be a solution to me
> 
> for me it does not make sense to display a warning at all,
> just talk about it in the upgrade guide for v2,
> but stop spamming with deprecation notices when the end user don't care about,
> and can't do nothing about it because those deprecation notices are from node_modules.

It would be nice to at least have an `quiet` and `quiet-deps` options in node-sass that ignores node_modules directory as I'm guessing for 99% of the users they will be importing other peoples libraries that they don't have control over.  Some dev importing it will probably hit the deprecation warning and let that library's dev know about it.

**Comment by flachware** [comment]
I admit that this is a very tricky issue, but especially as a long-term Sass user I’m sorry to say that from my point of view the `math.div()` solution does not do justice to the claim of being 'syntactically awesome'.

What we have now:
```
@use "sass:math";

div {
  padding: $space-sm + 2px;
  margin: math.div($space-sm, 2); 
}
```
The problems I see here: a) it is confusing that one basic arithmetic operation is represented by a function call whereas the other three are not, b) the function requires knowledge about the module system, c) it is hard to read, especially in case there are multiple divisions in a single calculation that now require nested function calls, d) it is cumbersome to write.

Discarded alternatives:
```
div {
  padding: $space-sm + 2px;
  margin: $space-sm ÷ 2;
}

div {
  padding: ($space-sm + 2px);
  margin: ($space-sm / 2);
}

div {
  padding: $space-sm + 2px;
  margin: ($space-sm / 2);
}
```

I’m a bit surprised that some of the mentioned alternatives above were ruled out early and did not get any further consideration. 

`÷`: granted, a key combination is required to enter the division sign, which is not great. But having to type `@use "sass:math"; math.div(,)` for a simple division isn’t great either – plus the cons mentioned above. However, I do see a legibility issue here, ÷ looks arguably a bit close to + on low-res screens.

`()`: fencing all math inside parentheses was ruled out as being a 'too severe' change. I fail to see the big disadvantage here, existing Sass code has to be migrated either way and new projects can adopt any syntax. This approach would be consistent, more concise than any other solution, does not require knowledge about the module system, and it would be universal (parentheses are used in maths anyway). Deprecating an operator is a severe change in any case.

And then there is the option to only fence calculations that contain the division sign with parentheses, like it is done in Stylus (and probably other preprocessors). This is a bit inconsistent, but that’s really the only con here. Maybe I’m missing something here, but either of the 'naked parentheses' options looks more promising to me than the concluded solution.

I am aware that it is technically 'too late' for suggestions, but I believe it’s never too late for improvements, so I made my case for a simpler solution.

**Comment by nex3** [comment]
@kerryj89 https://github.com/sass/sass/issues/3065 is tracking adding those flags to Node Sass. It's difficult to square them with the current importer API, so unfortunately we have to wait until the new importer API is released—but fortunately we're actively working on that right now.

@flachware I agree that the syntax is not as great as we'd like. If it were up to me, I would have pushed the CSS working group to avoid expanding the use of `/` as a separator so that we could go on using our old heuristic that mostly treated `/` as division. Unfortunately, the single strictest design guideline we have to follow is compatibility with CSS. It even takes precedence over beautiful syntax when necessary.

Using parentheses to tag division has a number of serious downsides:

* It's only *subtly* different than the old `/` heuristic. This makes it hard for both humans and automated tools to determine whether a given stylesheet was intentionally written to support `/`-as-separator, which makes migration harder to accomplish and deprecation warnings more likely to be missing or wrong.

* It makes syntax parsing context-sensitive. Users expect to be able to copy an expression from one context to another and have it work the same way, and when you break that invariant people start getting really confused. This is particularly problematic for newer users who are more likely to be copy-pasting snippets from StackOverflow or exisiting code.

* The way parentheses in Sass interact with lists and maps is already a point of confusion for a lot of users, and adding *more* meaning to them will compound this problem. It also opens up a bunch of questions without obviously-correct answers, such as how to parse `(1/2 1/3)` or `("foo": 1/2)`.

I will say that one of the major features we're working on this year is first-class calc (#818), which *will* provide a context in which `/` can be used as division again. It's more verbose and has more knock-on differences than the old ability to use `/` anywhere-ish, but it won't require a separate `@use` rule.

**Comment by essenmitsosse** [comment]
Ok, hear me out: **Feature flag.** 

Change the default behaviour to `math.div` but add a feature flag (on a per file basis or whatever) to allow the old way, maybe while disallowing the new `/`-as-separator interpretation. This would at least make migrating a bit less of a hassle for old projects, while not breaking the rule to have CSS be valid Sass (since CSS wouldn't have the feature flag).

**Comment by essenmitsosse** [comment]
Ok, hear me out: **Feature flag.** 

Change the default behaviour to `math.div` but add a feature flag (on a per file basis or whatever) to allow the old way, maybe while disallowing the new `/`-as-separator interpretation. This would at least make migrating a bit less of a hassle for old projects, while not breaking the rule to have CSS be valid Sass (since CSS wouldn't have the feature flag).

**Comment by flachware** [comment]
@nex3 thank you for your response. Up front, none of 'my' proposed alternatives (they are actually from you and @mirisuzanne) are in conflict with the compatibility with CSS.

I beg to differ, I think in the context of a programming language parentheses are not a subtlety, they are either there or not, for both humans and parsers.

Re your second point, I fail to see how `math.div()` solves this better than naked parentheses. If I copy & paste old Sass code it will be invalid, if I copy & paste pure CSS it will be valid in both cases.

In my understanding `()` would mean 'this is a calculation' (so it’s not my intention to tag divisions but to declare calculations). Think of `()` the same way as you would for the upcoming first-class `calc()` except that it would not fall back to the CSS `calc()`. In this sense `(1/2 1/3)` and `("foo": 1/2)` are not ambiguous but clearly invalid calculations and would throw a compiler error.

It’s good to hear that you will be working on the first-class `calc()` soon, because this is in line with my strategy to update the deprecated `/` directly to it, mainly for reasons of readability. But at this point I am wondering again, why is `calc($space-sm / 2)` feasible and `($space-sm / 2)` not? Would it be a possibility to make `()` a shorthand of the first-class `calc()`?

I guess the `math.div()` solution crosses the line because it makes divisions even more complicated than in CSS itself, I think that’s the bar here. The whole point of Sass is to make stylesheets easier to handle, if part of the syntax is more complicated it should also be more powerful. So it’s my understanding that the Sass syntax has to be more elegant and concise than the CSS syntax generally.

**Comment by mirisuzanne** [comment]
We all agree this is a frustrating situation, and that `math.div()` is not an ideal solution. But it's a very much needed temporary fix while we work on the larger, more elegant solution. Even if the previous behavior wasn't breaking _your_ site, it was breaking our core promise of CSS compatibility, and it had to be changed ASAP. That change was discussed in depth last year, and is now underway.

- As a policy, we avoid flags that would make the same file behave differently in different contexts. I don't expect that to change any time soon.
- The first-class `calc()` function will address the primary concerns raised here, while also providing a clarity of syntax that makes it much more future-proof. Parenthesis alone do not provide that assurance. Other half-steps in that direction only distract from the more robust solution.

We spent a lot of time considering all the tradeoffs, before we came up with a plan that gets us there as safely as possible. That all happened here, in the open, last year. We know it's not a painless transition, and we know that `math.div()` is not the best we can do long-term -- but it does address the immediate problem, while we work on a more elegant solution. And it avoids a lot of potential issues along the way. That discussion about the best path forward happened here in the open last year, and the resulting plan is now actively underway. Re-hashing the first-step of that plan, after-the-fact, doesn't actually help us move forward. 

At this point, we all just want to get first-class `calc()` rolled out as soon as possible – and that's where the focus will remain.

**Comment by whaaaley** [comment]
I've been using Sass for over a decade and never had issues with division or `/` as a separator. I think the deprecation should have held off until `calc()` was rolled out. The `/` was the status quo since the start and by deprecating it before an actual usable alternative is released seems a bit silly to me. I would bet real money that this preemptive deprecation will cause more headaches than headaches from "breaking css compatibility".

Why warn people to update everything to `math.div` when `calc()` is just around the corner? That's two massive updates vs just one. Especially if `calc()` will actually be released "asap".

I'll be locking my Sass version to the previous patch until `calc()` is released and would recommend everyone else do as well to avoid making multiple migrations. Good luck guys.

Edit: Version `1.32.13` for those wondering.

**Comment by moleCuleFFF** [comment]
It starts ....

**Comment by moleCuleFFF** [comment]
I think I'll be moving to https://stylus-lang.com/ if not just for the improved syntax, I'll leave this here for like minded developers.

FYI I am in no way affiliated or represent the stylus community but think people should be aware there are options out there.

**Comment by nex3** [comment]
@whaaaley I think it's hyperbolic to say that `math.div()` isn't usable. It works the same way as many other mathematical operations in Sass, and while division is somewhat more common than `math.pow()` or `math.min()` the extra four characters and a `@use` rule are not as onerous as you're making them out to be. As we continue to expand Sass's core library and work towards deprecating all top-level Sass functions, it's a reasonable expectation that *most* Sass files will begin with one or two core-library `@use`s.

We're deprecating this now for several reasons. First, we generally try to deprecate old behavior as soon as we know it's going to be removed at some point in the future. This gives users as much time as possible to move away from it. Second, it's not obvious that `calc()` is the ideal solution for all circumstances anyway. While it is terser, it raises the possibility that a given `calc()` expression will accidentally return a Calculation object rather than a number in some circumstances and then be incompatible with other Sass operations.

Ultimately, as Miriam said, this discussion has already been done and settled in public, and the decision has been made. There's not much point relitigating it now. If the deprecation warnings are bothering you, my primary recommendation is to run the migrator, but if you're unwilling to do that for whatever reason running in `--quiet` mode is certainly better than pinning to an old version and losing out on compatibility with libraries and new CSS features.

**Comment by whaaaley** [comment]
@nex3 I can agree that `math.div` is fine for many use cases, but in my case I have many color conversion functions and other algorithm functions that use a LOT of division and converting those to `math.div` is a readability nightmare. So I think we can just agree to disagree here.

As for "relitigating"... I think it's a good thing to keep in mind that many people don't follow the development of this project closely and the deprecation warning will likely be the first time people learn about this change. I imagine in the coming weeks as people update their dependencies more people will find their way to this issue wondering what happened.

I think it's better to keep an open mind to suggestions given how critical and common the division operator is.

Anyway, I'm not trying to dictate how to run your project. I'm just trying to offer some perspective. I've loved using Sass for as long as I've been using it and I like many of the changes that have been made over the years.

Edit: And for whatever it's worth, I do agree with @flachware's suggestion. I think fencing math in parens would be a great solution and deserves ~~some~~ _more_ consideration. I think it's 100x less intrusive than slapping `math.div` in the middle of your equation.

**Comment by Muffinman** [comment]
As many others will after me, I found this thread after seeing a warning after updating one of our projects.

I have to say this change is _really_ surprising to me. Sass is such a great tool for developers exactly because it generates readable, flexible CSS with very little cognitive overhead for the developer.

I feel like the DX deterioration of this change has been thoroughly overlooked. I understand you have no control over the CSS working group adding these additional uses for `/`, but I still feel that it would have been 100x better to find a new way of defining these `grid-row-start` and `grid-row-end` directives rather than throwing out perfectly elegant syntax (along with introducing the issue of 1000s of old projects that now require manually updating).

In fact, as a developer, I don't think I would ever write `grid-row: 1 / 3;` in sass and expect it to work. I would most likely write separate directives or try to wrap in quotes. For how infrequently I write this compared to using a division operator it's really a no-brainer for me.

Another thought here (which may not be of concern for you), is that developers will seriously think twice about using sass on new projects given the possibility of other such drastic changes to syntax in the future. This is especially true given the current momentum of Tailwind and the like.

/2p

**Comment by lehni** [comment]
100% agreeing with everybody else here. Our Vue / Nuxt based project spits out 100s of these warnings after the upgrade, slowing down the build **a lot**. Converting the full code-base will be far from trivial, resulting in much less readable code. It's a bizarre change of behavior that was rolled out with a minor version change. We downgraded back to the last version that doesn't warn about this and wait for the dust to settle. I have strong doubts that this is the final decision in this discussion.

**Comment by lehni** [comment]
Also, why was @moleCuleFFF's comment marked as spam? They point out that another library has found a different way to handle this operator ambiguity, and I think that's valuable information.

**Comment by nex3** [comment]
@whaaaley We go to great pains to make our design and decision-making process as open and participatory as possible, and to publicize major decisions on the Sass blog and Twitter account. You're right that some people don't follow it closely, but part of what that means is that those people have a weaker ground for objecting to the decisions that get made when they *could have* participated in the process. The Sass team is very busy, and we don't have infinite time and energy to go round and round on old decisions. Part of the point of having such a deliberative process in the first place is so that we can lock in decisions and move on.

That said, because this is a hot-button issue, I'll walk through some of our design reasoning in detail. First: **matching CSS syntax is non-negotiable**. It is *the* core syntactic design goal for Sass. This gets at @Muffinman's point: even if *you personally* would expect `grid-row: 1 / 3` to perform division, in CSS it unambiguously *does not do so*. This focus on compatibility isn't mere idealism: it's crucial for onboarding new authors to Sass, because they invariably know CSS better than they do Sass; and it's crucial for avoiding even more disruptive breaking changes in the future, because the CSS working group won't hesitate to add new features that re-use any syntax they've claimed as their own. So far we've been able to avoid that with `/` because the `font` shorthand is optional, we've hacked in special support for the color syntax, and grid support isn't something that often needs to be stored in a variable—that buys us time to do a proper deprecation process and give everyone time to upgrade. But if we just stick with the old behavior, it would be entirely possible for CSS to introduce a new feature that unavoidably uses it in a way that breaks Sass's heuristics and we'd have to do the same dance but in a matter of months rather than years.

Second, relating specifically to the possibility of fencing division with parens, I went into some detail [above](https://github.com/sass/sass/issues/2565#issuecomment-852563552) about the downsides of that approach. There's another general principle at work though that's worth mentioning: **verbosity is less bad than surprise**. Ideally Sass would be terse *and* unsurprising, but when our hand is forced (as it is here), it's better to err on the side of heavyweight-but-obvious syntax than lightweight-but-surprising. (I acknowledge that for people coming from other programming languages, `/` not meaning division is *itself* surprising, but we need to prioritize the surprise for people coming from a CSS background and we can't make Sass match other programming languages' syntax without violating CSS compatibility anyway.)

As @mirisuzanne pointed out, first-class `calc()` (#2186) will provide a similar solution for providing a context in which `/` means division, while being more explicit and less overloaded than parentheses. Will this be satisfyingly terse for you all? If so, I'm prepared to prioritize its development to minimize the amount of time that `math.div()` is the only division solution that doesn't produce a warning.

@lehni 

> Our Vue / Nuxt based project spits out 100s of these warnings after the upgrade, slowing down the build **a lot**.

Upgrade to Dart Sass 1.34.0; it limits the number of repeated deprecation warnings that are printed by default to 5.

> Also, why was @moleCuleFFF's comment marked as spam? They point out that another library has found a different way to handle this operator ambiguity, and I think that's valuable information.

"Don't use Sass, use this other stylesheet preprocessor!" is not an appropriate thing to post in the Sass issue tracker.

**Comment by whaaaley** [comment]
@nex3 I appreciate you taking the time to explain this in more detail. I think it helps everyone here understand if they weren't here from the very beginning.

> Will this be satisfyingly terse for you all? If so, I'm prepared to prioritize its development to minimize the amount of time that `math.div()` is the only division solution that doesn't produce a warning.

For me? Absolutely!  `calc` is readable by my standards. My main issue with `math.div` was readability more so than backwards compatibility. Especially because there's a migration tool anyway. As long as `calc` can be used in variables and the math inside can be still be reduced to it's smallest form then I'll be a happy dev. 👍

**Comment by lk77** [comment]
I think the real issue is the deprecation warning,
we should not have to rollback to prior sass versions to avoid those deprecation warnings.

**Comment by nex3** [comment]
All versions of Sass have `--quiet` and `--quiet-deps` CLI options, the Dart API has these available as API options as well, and we're hard at work making them available as API options in JS. We also busted our butts to get out a release just *one day* after the initial deprecation warnings launched which limited the build to five deprecation warnings per type. We're doing our best!

**Comment by moleCuleFFF** [comment]
@nex3 I didn't say don't use sass use this I said I'm using this because I think they solved the problem the correct way, sorry but you are WAY too triggered by a post promoting competition, if sass is better than stylus you shouldn't be worried.

Educating other devs on alternatives is perfectly reasonale in this space, SASS is OS it's not yours to own, I could fork it whenever.

Where is the community guidelines regarding promoting other technologies ?

You really aren't being 

a) Very helpful
b) Very nice
c) very OS

Saying you didn't follow the issue so don't get an opinion is un-constrictive and unrealistic, I do not spend my time concentrating on how every tool is going to work going forward, you should expect backlash on an issue like this and handle it more approporiately by maybe re-opening the issue now it's garnered further public attention and asking devs for solutions, shutting people down isn't very OS.

Unless ur expecting me to keep tabs on the 20,000 odd node packages in my applications at all times for upcoming deprecation warnings so I can have an opinion, it's an unrealistic expectation.

**Comment by lk77** [comment]
edit: was replying to a deleted comment

well to be fair, i think sass is a core package of the ecosystem, worth looking at, we are not talking about some random package, deep inside the dependency tree.

**Comment by lehni** [comment]
@nex3 thank you for the explanations. Regarding your suggestion:

> Upgrade to Dart Sass 1.34.0; it limits the number of repeated deprecation warnings that are printed by default to 5.

We're using Sass in a large project with many single-file components, many of them having their own style definitions. I think it's something about this setup that makes it not respect that setting globally and spit out an insane amount of deprecation warnings, slowing down the build a lot.

Prioritising `calc()` in sass sounds like a great way forward, BTW. I think we will wait with upgrading until that is available, as it will help with readability.

**Comment by Nyatra** [comment]
Just updated, found the deprecation warnings, read this thread, and downgraded to avoid the deprecation messages.

Just because we aren't following the blogs and social medias of the tools and libraries we are using doesn't mean we have "weaker ground for objecting". The majority of users are not going to follow everything, it's just unreasonable. I already follow way more than I would like to. Deprecation warnings and errors in code are when people will really find out about changes.

The problem seems to stem from new uses of "/" as a separator in native CSS, which should just be edge cases of those new additions in sass and shouldn't be affecting the use of "/" as division everywhere else. If "math.div()" or "calc()" was just the solution to those edge cases, this wouldn't be a big deal.

For global usage, "math.div()" is too verbose and also makes for very unreadable code.
```
width: math.div(math.div($var1, $var2) * $var3, $var4)
```

If it's going to be a global change, then naked parenthesis/fencing should have been the direction taken with this.
```
width: ($var1 / $var2 * $var3 / $var4)
```

What happens if they start using "*" to indicate something in CSS values? Add "math.mul()"?
```
width: math.div(math.mul(math.div($var1, $var2), $var3), $var4)
```

**Comment by nex3** [comment]
**Update:** I figured out a way to make the `quietDeps` flag work with the JS API. Dart Sass 1.35.0 is [releasing now](https://github.com/sass/dart-sass/actions/runs/937639036) with that flag added, which should make it possible to unpin your versions without being deluged by warnings you can't control. Just pass `quietDeps: true` wherever you pass in options to Dart Sass!

@wulf2468 I've explained numerous times throughout this thread why failing to support plain CSS is not an option and the numerous downsides of naked parenthesis fencing. I'm not going to argue further about it.

**For everyone:** I'd like to avoid locking this issue, but I'm not interested in going around in circles about things that have already been explained in detail and discussed to death. Please refrain from posting more unless you have something actually new to bring to the conversation.

**Comment by Nyatra** [comment]
As people update sass, more and more people are going to be coming in here to DISCUSS the changes and try to come up with alternatives that don't require them to rewrite and obfuscate all their existing code to support something they and the majority of users aren't going to be using.

A blacklist of CSS properties/functions could be added that cannot use "/" as division. For those, "/" would only be treated as a separator and "math.div()" or "calc()" would have to be used to do division. Plain CSS would be supported and all existing code would continue to work as it does now.

**Comment by nex3** [comment]
> A blacklist of CSS properties/functions could be added that cannot use "/" as division. For those, "/" would only be treated as a separator and "math.div()" or "calc()" would have to be used to do division. Plain CSS would be supported and all existing code would continue to work as it does now.

This runs into the same issue of making the same Sass expression do different things in different contexts. It's really surprising if `foo(<expression>)` does one thing and `$var: <expression>; foo($var)` does something totally different. As I explained above, when we're faced with a difficult choice we prioritize unsurprising behavior over terse behavior.

**Comment by Nyatra** [comment]
I would say `width: 100px + $var;`, `width: 100px - $var;`, `width: 100px * $var;` working, but `width: 100px / $var` not working is far more surprising. Even with new properties and functions being added that use "/" as a separator, division will be used far more often than those would.

**Comment by nex3** [comment]
Yes, that will be surprising. We're not *happy* to get rid of `/`-as-division, and we acknowledge that this is a bad situation all around. But we have a commitment to CSS compatibility, so our hands are tied.

**Comment by Muffinman** [comment]
Thank you for your detailed responses. It's clear that this is a decision that has been made and will not be changed, but I do feel it warrants some further discussion on if anything further could be done to appease developers who may ongoing dread having to go back to work on any old Sass project.

Is there no possibility of implementing a config/flag to make this behaviour developer choice? I feel this would be the best of both worlds. Obviously the default can be left as your new `math.div()` going forward, but it's really just to try to avoid developer friction upgrading the many thousands of old projects using sass.

@mirisuzanne above states that you try to avoid adding flags for behaviour wherever possible, but I really think that a large scale syntax change like this may be a justified case?

This is not just about being lazy and not wanting to update our own code or write a few extra characters; if including third party npm Sass files (e.g. [Sweetalert](https://github.com/sweetalert2/sweetalert2)) it then becomes a far bigger headache to update, especially if the library has not been updated for the new syntax. It could easily make what would otherwise be hour of work on an old project to more like a day.

**Comment by lk77** [comment]
if those project are really old/unmaintained, i don't see any reason to update sass anyway

**Comment by nex3** [comment]
We're not going to indefinitely support a flag that enables a syntax that's incompatible with CSS. As Miriam says, behavioral flags are disruptive to the ecosystem, confusing to users, and onerous for the implementation.

The reason we have a long deprecation period is to give the ecosystem a long runway to upgrade before we actually make a breaking change. In this case it's likely to be *years* before `/`-as-division is gone for good. That's plenty of time for everyone to move over to the new behavior.

**Comment by luigimannoni** [comment]
I think there's not much to add apart the solution already proposed above, so I'll just reiterate what are my preferences on what I've read so far down here. It's just a repetition of what said with a bit of reasoning.

Strongly oppose to remove `/` as division while keeping the other operators working, I can safely assume that's illegal speaking of coding in general not only the quickest way to induce blood boiling in a Dev.
Personally it makes me want to open a petition on change.org, call the police, start a worldwide movement and ultimately stick to a legacy version of sass and never update again. For the last point I can see it becoming a truth in no time.

Strongly oppose to use verbose stuff such as math.div, math.sum, etc. Sass should be quick to write, quick to read for any programmer, increasing verbosity makes code reviews 20 times more difficult for simple calculations. `sum(div(100%, $columns), mul($grid-gutter, dif($column, 1)))` takes a bit of time to decipher in this way, while it is an instant in its usual format. Operators are basic maths taught in schools, everyone can read them efficiently.

I support the idea to use parenthesis/fencing, which is also sometimes enforced by some linters while mixing /, * and +, - operators. `width: (100px / 2)` should be always safe to type and the concept is easier to grasp and quicker to type. Also keeps code readable.
Stylelint can readapt in no time for this syntax and it's far easier to implement an autofix which adds all the missing parenthesis rather than transforming operators into comma separated functions. Old projects will be a breeze to convert into the new syntax using `stylelint --fix`

I also support the idea to start teaching back to frontend devs a bit of native CSS by start enforcing use `calc()` for all math calculations as it's meant by CSS standard. I've been a slave and dependent of Sass that I am often under the impression that `calc()` is a hacky thing rather than a legit function to use as it is intended to be used.
Sass can optimise the output by solving `calc(100px / 2)` into a final `50px` so that there's no calc overhead in the final stylesheet, mixed units can retain the calc function as well as making feel the native `calc()` function more legit. 
I feel it's a win-win situation.

Also will likely solve the other outstanding issue of using the `#{$var}` "quirk" to transform variables into number within the calc function.

Can parenthesis `()` be a shorthand for `calc()`? I think both should need to behave the same way at the end

**Comment by khamer** [comment]
Just expanding on my earlier comment:

```scss
@use "sass:math" as *;

@function divide($a, $b) {
    @if function-exists(div) {
        @return div($a, $b);
    } @else {
        @return $a/$b;
    }
}
```

I tested back as far as 1.28 on dart sass and tested node-sass 6.0.0 (libsass 3.5.5.) The use statement shows up within the compiled result and while it's an unrecognized rule, I believe that this will function properly and not throw any deprecation notices. Using the migrator or the new flag to get a single deprecation notice are both great - but just wanted to share this expanded temporary solution if you're writing a codebase that you want to run on dart-sass before and after 1.33. If you only have to support 1.34+, then I wouldn't bother with this.

**Comment by moleCuleFFF** [comment]
Why are we minimizing all the comments we don't agree with ? Stop silencing the open source community.

**Comment by moleCuleFFF** [comment]
> 
> 
> I think there's not much to add apart the solution already proposed above, so I'll just reiterate what are my preferences on what I've read so far down here. It's just a repetition of what said with a bit of reasoning.
> 
> Strongly oppose to remove `/` as division while keeping the other operators working, I can safely assume that's illegal speaking of coding in general not only the quickest way to induce blood boiling in a Dev.
> Personally it makes me want to open a petition on change.org, call the police, start a worldwide movement and ultimately stick to a legacy version of sass and never update again. For the last point I can see it becoming a truth in no time.
> 
> Strongly oppose to use verbose stuff such as math.div, math.sum, etc. Sass should be quick to write, quick to read for any programmer, increasing verbosity makes code reviews 20 times more difficult for simple calculations. `sum(div(100%, $columns), mul($grid-gutter, dif($column, 1)))` takes a bit of time to decipher in this way, while it is an instant in its usual format. Operators are basic maths taught in schools, everyone can read them efficiently.
> 
> I support the idea to use parenthesis/fencing, which is also sometimes enforced by some linters while mixing /, * and +, - operators. `width: (100px / 2)` should be always safe to type and the concept is easier to grasp and quicker to type. Also keeps code readable.
> Stylelint can readapt in no time for this syntax and it's far easier to implement an autofix which adds all the missing parenthesis rather than transforming operators into comma separated functions. Old projects will be a breeze to convert into the new syntax using `stylelint --fix`
> 
> I also support the idea to start teaching back to frontend devs a bit of native CSS by start enforcing use `calc()` for all math calculations as it's meant by CSS standard. I've been a slave and dependent of Sass that I am often under the impression that `calc()` is a hacky thing rather than a legit function to use as it is intended to be used.
> Sass can optimise the output by solving `calc(100px / 2)` into a final `50px` so that there's no calc overhead in the final stylesheet, mixed units can retain the calc function as well as making feel the native `calc()` function more legit.
> I feel it's a win-win situation.
> 
> Also will likely solve the other outstanding issue of using the `#{$var}` "quirk" to transform variables into number within the calc function.
> 
> Can parenthesis `()` be a shorthand for `calc()`? I think both should need to behave the same way at the end

Is a really good response.

**Comment by moleCuleFFF** [comment]
I think as a community we should fork sass and run a seperate repo for it. Google are riding ruffshod over community decisions here.

Open source is not "Open an issue for a few months consultation within Google not advertising the change to the community vocally and then shutting down discussion about it post implementation of the deprecation warning"

The deprecation warning should have been the start of the discussion.

This comment will be minimised, watch this space.

**Comment by moleCuleFFF** [comment]
I tried to post the use of Stylus as a solution as it already implements the fencing as a lot of devs here are asking for .... silenced.

Competition should be a good thing ! Maybe sass could look at the Stylus implementation and take some inspiration, but no they just wanna cancel my voice and prevent othere devs finding solutions outside their control.

Even the CSS guys reccomended fencing.
>We did vacate the naked-parenthesis space for y'all, and I'm willing to defend that line, so if you want to reserve "within ()" as the space for safe free-form math, I think you can safely do so. Or, of course, just use "calc()". ^_^

Really not cool, Google don't own SASS.

**Comment by nhhockeyplayer** [comment]
build is blocked in my camp

**Comment by nhhockeyplayer** [comment]
build blocked in my camp

so this change takes out the whole ecosystem
how do we get past the packages its compiling against under node_modules
we wait til new ones are released?

who's driving here?

the whole software plateau has been taken out

I made recommended changes
and cannot get past node_module dependencies that are still using the slash

???

**Comment by nhhockeyplayer** [comment]
>We downgraded back to the last version that doesn't warn about this and wait for the dust to settle. I have strong doubts that this is the final decision in this discussion.

can someone share which rev to downgrade to? the package(s) and rev(s)

this is outrageous

the jest packages took out the whole build ecosystem in late May

now theres a webpack one too

why do people keep breaking everything what the hell is going on?

**Comment by denik1981** [comment]
Just to add this is issue because I think is fully related. 

I was having a unnecessary reference call to **Math.div()** inside a native CSS **calc( )** funtion. 
Despite it was useless, it was working well before and now the compiler seems to stopped parsing any Math.div() it founds inside a calc() function. I was able to get compiled any other use of Math.div except for this one inside calc.

So to add, SASS not only is not looking for a slash bar anymore but it is also is skipping parsing any Math.div it finds inside a calc() CSS function.

**Comment by lehni** [comment]
I'm aware that this discussion was getting heated and the arguments for the current decision had to be repeated and that was getting annoying, but repeatedly minimising comments and then blocking people who didn't use any kind of strong language in their posts, without any warning or heads up before, seems excessive. This will almost certainly put other people off. Please reconsider how you handle this discussion…

**Comment by luigimannoni** [comment]
I am not that much fussed about my comment being minimised, I just repeated what everyone else said already above and made a point. I understand the reasons of why the slash needs to go, but I don't understand who took the final decision to use that convoluted way to make divisions and who agreed on that since most of the feedback on this issue seem to steer in another direction.
Sadly it came late as the damage has been already merged into the main branch of the project, I mean first result on Google is already suggesting to downgrade Sass to circumvent the deprecation error.

I am more concerned about how this decision has been taken and how a change of this scale of impact has been not communicated to the community in the more appropriate way, #3079 raises a valid point.

**Comment by lk77** [comment]
yeah i've rolled back to prevent deprecation errors to 1.33,
and i think it's the best solution for now, even if there is a 5 message limit now,
i can't put something in production with deprecation errors or warnings in it.
Builds logs needs to be perfectly clean.
It's the only package that i know that does that, deprecation should be noted in the changelog, not printed in the console.

**Comment by cyraid** [comment]
Would a 'math block' denoted by a special processing work? In bash $(( 1 + 1 )) is what allows you to do math. Would something similar be an option for Sass?

**Comment by lk77** [comment]
@cyraid already discussed, and it would not be an option, for css compatibility it seems, and there is the native css calc() function that does that too.

The most adopted solution in the community is to replace division by multiplication, instead of using math.div, it seems that math.div should be avoided if compatibility with libsass matters.

**Comment by cyraid** [comment]
> @cyraid already discussed, and it would not be an option, for css compatibility it seems, and there is the native css calc() function that does that too.
> 
> The most adopted solution in the community is to replace division by multiplication, instead of using math.div, it seems that math.div should be avoided if compatibility with libsass matters.

Calc is browser side. But the new super calc that's coming should at least reduce to minimum. If it can also remove the calc if it's not necessary that basically is your golden bullet.

**Comment by mirisuzanne** [comment]
@cyraid That is the planned behavior of first-class `calc()`: https://sass-lang.com/blog/request-for-comments-first-class-calc

**Comment by nex3** [comment]
> Just to add this is issue because I think is fully related.
> 
> I was having a unnecessary reference call to **Math.div()** inside a native CSS **calc( )** funtion.
> Despite it was useless, it was working well before and now the compiler seems to stopped parsing any Math.div() it founds inside a calc() function. I was able to get compiled any other use of Math.div except for this one inside calc.
> 
> So to add, SASS not only is not looking for a slash bar anymore but it is also is skipping parsing any Math.div it finds inside a calc() CSS function.

@denik1981 Sass (currently) never parses anything within `calc()`. It's treated as a kind of string quote. Nothing about the `/` deprecation changed this behavior. See [the documentation](https://sass-lang.com/documentation/syntax/special-functions) for details.

> I'm aware that this discussion was getting heated and the arguments for the current decision had to be repeated and that was getting annoying, but repeatedly minimising comments and then blocking people who didn't use any kind of strong language in their posts, without any warning or heads up before, seems excessive. This will almost certainly put other people off. Please reconsider how you handle this discussion…

@lehni Posting four times in a row and reposting comments that had been minimized is combative and clearly against the guidelines I'd laid out for continuing to contribute to this thread.

> I am more concerned about how this decision has been taken and how a change of this scale of impact has been not communicated to the community in the more appropriate way, #3079 raises a valid point.

@luigimannoni As I described above, we posted about this change on the Sass blog and the Sass twitter account, as well as making all these changes fully in the open on GitHub. #3079 will help us broadcast even broader, but it's certainly not accurate to say that this change was "not [appropriately] communicated to the community".

> yeah i've rolled back to prevent deprecation errors to 1.33,
> and i think it's the best solution for now, even if there is a 5 message limit now,
> i can't put something in production with deprecation errors or warnings in it.
> Builds logs needs to be perfectly clean.
> It's the only package that i know that does that, deprecation should be noted in the changelog, not printed in the console.

@lk77 Deprecation warnings are the only reliable way to make sure everyone sees a change coming in time to prepare. A changelog entry isn't going to be read and acted on any more than the blog post we made requesting comments on this change.

It's silly to say that you "can't" put something into production with deprecation warnings. That's what warnings are for: providing notifications that things need to change *without* blocking deployment. If you choose to be warning-clean, we've provided `math.div()` as a way to make that happen, but downgrading is only going to lead to further compatibility woes down the line with CSS and with future Sass versions.

**Comment by Nyatra** [comment]
Seeing as the majority of your community are just installing sass through a package manager, probably 90%+ of the user base have not seen this change coming. A warning on compile or on package install, with a message like, "/ as division in sass will be changing due to increased usage of / as a separator in native CSS. Visit so-and-so GitHub issue to discuss the future of division in sass."

Regardless of whether the community saw it or not, it seems like there was no ability for the community to actually make a say in how this was even handled, since you "discussed this with a crack team of Sass users at Google, and we came up with the following plan".

As for the deprecation warnings, most people are downgrading, because not only does it spit out more than 1 warning, but each of them are also a paragraph long. It completely hides anything else going on in the output. A one line warning message like, "/ as division is deprecated and will be removed in a future version of sass. Visit so-and-so GitHub issue for more information." would suffice and get most people to not downgrade. If needed, an "upgrade hinting" flag could be added to activate the current functionality of showing each location of where deprecated features are currently being used for upgrade purposes.

Also, if you do put links into warnings, consider shortening them with a URL shortener to make the warnings shorter.

**Comment by cyraid** [comment]
> @cyraid That is the planned behavior of first-class `calc()`: https://sass-lang.com/blog/request-for-comments-first-class-calc

Yup! Looking forward to that. :)

**Comment by nex3** [comment]
@wulf2468 It's important that warnings provide a clear path forwards if at all possible. Very terse warnings are still annoying for people who already know about them (albeit less so), *and* they're extra confusing for people who don't have any context yet. We will almost always lean our design decisions towards working best for people who have the least context, because they're the least able to find and apply alternatives.

As has been mentioned above, we are planning to add a post-install message that solicits feedback about upcoming features.

**Comment by Akxe** [comment]
God, this is a bad change; why the universally used symbol should change its meaning in only a single one "language"?!

Would you please give us at least the option to toggle between default for "/" to be either division or list separator? I would strongly prefer "/" to be division as lists are, to me, so much more uncommon than simple division.

-----

I am certain that if this change rolls out to the wild a "new" language will appear that will only change this...

PS: Coming from Angular, this is the first time I saw this!

**Comment by Ashlyn-IO** [comment]
I dont understand why you need to get rid of / it breaks so many things. Saas will end up being its own convoluted language instead of being a tool to work with CSS. It really sucks when you have to google the documentation to do something as simple as a division. Is this really something the users want?

**Comment by Akxe** [comment]
@StefGraf they don’t care about user anymore. They seem to be only focused on their group and ideologies in it. Look how every message questioning the decision from the very beginning of this thread is marked as spam. Not a single word can be put against it. Guess I’ll be waiting for a **clone of sass reverting this greave mistake**.

**Comment by nex3** [comment]
As a reminder, we're keeping this thread unlocked specifically to allow *productive* discussion. Comments to the effect of "I don't like the decision you've made" will continue to be marked as spam.

**Comment by cyraid** [comment]
Hey guys, just a reminder of the awesome super `calc()` that's coming.. This will solve what you need, it'll boil down to using without `calc()` if it can. So anywhere you're using them you can slap in a `calc()`.. Wanna know another bonus of doing that?

Let's say you have:
```scss
$a: calc(4px * 2);
$b: calc($a / 2);
```

So `$a` should be just `8px`, and `$b` will end up being `4px`! Just like we wanted. Also, the huge added benefit? Let's say you want:

```scss
$spacing: 4px;
$width: calc(100% - ($spacing * 2));
```

It should turn into a browser calc:
```scss
$width: calc(100% - 8px);
```

**Comment by rolandoff** [comment]
Hey guys.
In the project that I'm currently working on, we update our dependencies and now we're having this issue where we get this error at build time: 

`512:7  ✖  Expected modern color-function notation  color-function-notation`

So this is happening when doing this in one of our sass files: 

`color: rgba($primary-color, 0.5);`

The problem comes now if we try to update to the modern notation, which will be something like this: 

`color: rgb($primary-color / 50%);`

Becase for SASS this is a division... So no idea how to fix this. Any clue?

**Comment by Awjin** [comment]
@rolandoff You can use the [`list.slash` function](https://sass-lang.com/documentation/modules/list#slash) to create a slash separated list: `color: rgba(list.slash($primary-color, 0.5))`.

**Comment by rolandoff** [comment]
> @rolandoff You can use the [`list.slash` function](https://sass-lang.com/documentation/modules/list#slash) to create a slash separated list: `color: rgba(list.slash($primary-color, 0.5))`.

Hey! Thanks for your answer. 
I just tried out and it didn't work. This is the error that I'm getting: 
![image](https://user-images.githubusercontent.com/456184/140554760-2ac850c8-e884-4d47-b024-13f2045ebab2.png)

I really think there is an issue on using sass variables within this new notation.

**Comment by nex3** [comment]
@rolandoff The issue is that `rgba($primary-color, 5%)` wasn't CSS syntax to begin with. That's a special Sass-only syntax that's equivalent to `color.change($primary-color, $alpha: 5%)`. So `rgba($primary-color / 5%)` isn't expected to work because (for example) `rgba(red / 5%)` isn't valid CSS. You should just keep on using `rgba($primary-color, 5%)` without the slash.

**Comment by michahell** [comment]
I am getting compilation errors from `sass-loader@12.2.0` when using for instance `div.math($variable, 2);`

**Comment by markdascher** [comment]
Is first-class calc implemented now? Just trying to figure out the current situation by reading https://sass-lang.com/documentation/breaking-changes/slash-div which states:

> Alternatively, users can wrap division operations inside a calc() expression, which Sass will simplify to a single number.

That sounds like the opposite of what's mentioned [here](https://sass-lang.com/blog/request-for-comments-first-class-calc) "each calc() is totally opaque to Sass’s parser." So does that mean there's no need for math.div anymore?

**Comment by nex3** [comment]
Yes, first-class calc is implemented. (The best way to tell is to look at the "Compatibility" widget on [the documentation page for calcuations](https://sass-lang.com/documentation/values/calculations), which says it was added in 1.40.0.)

> That sounds like the opposite of what's mentioned [here](https://sass-lang.com/blog/request-for-comments-first-class-calc) "each calc() is totally opaque to Sass’s parser."

That sentence is describing the way `calc()` was historically parsed, not the new proposal described in that blog post.

> So does that mean there's no need for math.div anymore?

`math.div()` is still useful because it's _guaranteed_ to either return a number or throw an error. `calc()` may return a calculation object, which can't be used for other top-level arithmetic operations or various Sass functions.
